contracts/gateway/src/contract.rs

```diff
@@ -1,54 +1,125 @@
 #[cfg(not(feature = "library"))]
 use cosmwasm_std::entry_point;
 use cosmwasm_std::{Binary, Deps, DepsMut, Env, MessageInfo, Response};
+use std::fmt::Debug;
 
-use crate::contract::execute::Contract;
-use crate::{
-    msg::{ExecuteMsg, InstantiateMsg, QueryMsg},
-    state::{Config, CONFIG},
-};
+use crate::msg::{ExecuteMsg, InstantiateMsg, QueryMsg};
 
 mod execute;
 mod query;
 
 #[cfg_attr(not(feature = "library"), entry_point)]
 pub fn instantiate(
     deps: DepsMut,
-    _env: Env,
-    _info: MessageInfo,
+    env: Env,
+    info: MessageInfo,
     msg: InstantiateMsg,
 ) -> Result<Response, axelar_wasm_std::ContractError> {
-    let router = deps.api.addr_validate(&msg.router_address)?;
-    let verifier = deps.api.addr_validate(&msg.verifier_address)?;
-
-    CONFIG.save(deps.storage, &Config { verifier, router })?;
-
-    Ok(Response::new())
+    Ok(internal::instantiate(deps, env, info, msg)?)
 }
 
 #[cfg_attr(not(feature = "library"), entry_point)]
 pub fn execute(
     deps: DepsMut,
-    _env: Env,
+    env: Env,
     info: MessageInfo,
     msg: ExecuteMsg,
 ) -> Result<Response, axelar_wasm_std::ContractError> {
-    let mut contract = Contract::new(deps);
-    match msg {
-        ExecuteMsg::VerifyMessages(msgs) => contract.verify_messages(msgs),
-        ExecuteMsg::RouteMessages(msgs) => contract.route_messages(info.sender, msgs),
-    }
-    .map_err(axelar_wasm_std::ContractError::from)
+    Ok(internal::execute(deps, env, info, msg)?)
 }
 
 #[cfg_attr(not(feature = "library"), entry_point)]
 pub fn query(
     deps: Deps,
-    _env: Env,
+    env: Env,
     msg: QueryMsg,
 ) -> Result<Binary, axelar_wasm_std::ContractError> {
-    match msg {
-        QueryMsg::GetMessages { message_ids } => query::get_messages(deps, message_ids),
+    Ok(internal::query(deps, env, msg)?)
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("gateway contract config is missing")]
+    ConfigMissing,
+    #[error("invalid store access")]
+    InvalidStoreAccess,
+    #[error("failed to serialize the response")]
+    SerializeResponse,
+    #[error("batch contains duplicate message ids")]
+    DuplicateMessageIds,
+    #[error("could not query the verifier contract")]
+    QueryVerifier,
+    #[error("invalid address")]
+    InvalidAddress,
+}
+
+mod internal {
```

Comment:
What's the benefit of having this `internal` mod?

Reply:
implicit conversion of `Error` to `ContractError` with the help of `?`